File: ios/DisbandiOS/Views/Voice/VoiceLobbySheet.swift
import SwiftUI

/// Peek into a voice channel before you're heard in it: who's there, and one
/// big button to join. Muting first is a tap away, so nobody hears you arrive.
struct VoiceLobbySheet: View {
    let room: VoiceRoom
    var initialMembers: [VoiceMember] = []

    @Environment(VoiceSession.self) private var voice
    @Environment(\.dismiss) private var dismiss
    @State private var members: [VoiceMember] = []
    @State private var joinMuted = false
    @State private var joining = false

    var body: some View {
        VStack(spacing: 0) {
            VStack(spacing: 14) {
                AvatarCluster(members: members, size: 64)
                    .frame(height: 104)
                    .padding(.top, 28)

                VStack(spacing: 4) {
                    Text(room.title)
                        .font(.title2.weight(.bold))
                        .foregroundStyle(Brand.textPrimary)
                    Text(summary)
                        .font(.subheadline)
                        .foregroundStyle(Brand.textMuted)
                        .multilineTextAlignment(.center)
                }
                .padding(.horizontal, 24)
            }

            ScrollView {
                VStack(spacing: 0) {
                    ForEach(members) { member in
                        HStack(spacing: 12) {
                            AvatarView(url: member.profile?.avatarUrl, name: member.name, size: 36)
                            Text(member.name)
                                .font(.body.weight(.medium))
                                .foregroundStyle(Brand.textPrimary)
                            Spacer()
                            if member.deafened {
                                Image(systemName: "headphones.slash").foregroundStyle(Brand.dnd)
                            } else if member.muted {
                                Image(systemName: "mic.slash.fill").foregroundStyle(Brand.textMuted)
                            }
                        }
                        .padding(.vertical, 10)
                        if member.id != members.last?.id {
                            Divider().overlay(Brand.divider)
                        }
                    }
                }
                .padding(.horizontal, 24)
                .padding(.top, 18)
            }

            controls
        }
        .background(Brand.surface.ignoresSafeArea())
        .onAppear { members = initialMembers }
        .task { await poll() }
    }

    private var summary: String {
        switch members.count {
        case 0: return "No one's here yet — be the first."
        case 1: return "\(members[0].name) is hanging out"
        case 2: return "\(members[0].name) and \(members[1].name)"
        default: return "\(members[0].name), \(members[1].name) and \(members.count - 2) other\(members.count == 3 ? "" : "s")"
        }
    }

    private var controls: some View {
        HStack(spacing: 12) {
            Button {
                joinMuted.toggle()
                UISelectionFeedbackGenerator().selectionChanged()
            } label: {
                Image(systemName: joinMuted ? "mic.slash.fill" : "mic.fill")
                    .font(.system(size: 20, weight: .semibold))
                    .foregroundStyle(joinMuted ? Color.white : Brand.textPrimary)
                    .frame(width: 58, height: 58)
                    .background(joinMuted ? Brand.dnd : Brand.elevated, in: Circle())
            }
            .accessibilityLabel(joinMuted ? "Join muted" : "Join unmuted")

            Button {
                joining = true
                Task {
                    if joinMuted != voice.micMuted { voice.toggleMute() }
                    await voice.join(room)
                    joining = false
                    if voice.isIn(room.id) { dismiss() }
                }
            } label: {
                HStack(spacing: 8) {
                    if joining {
                        ProgressView().tint(.white)
                    } else {
                        Image(systemName: "waveform")
                    }
                    Text(joining ? "Joining…" : "Join Voice")
                }
                .font(.headline)
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity)
                .frame(height: 58)
                .background(Brand.online.gradient, in: Capsule())
            }
            .disabled(joining)
        }
        .padding(16)
        .background(.ultraThinMaterial)
    }

    private func poll() async {
        while !Task.isCancelled {
            if let loaded = try? await VoiceSession.fetchMembers(of: room) { members = loaded }
            try? await Task.sleep(nanoseconds: 4_000_000_000)
        }
    }
}

/// Up to four faces in an overlapping cluster, or a speaker glyph when empty.
struct AvatarCluster: View {
    let members: [VoiceMember]
    var size: CGFloat = 56

    var body: some View {
        let shown = Array(members.prefix(4))
        ZStack {
            if shown.isEmpty {
                Image(systemName: "speaker.wave.2.fill")
                    .font(.system(size: size * 0.5, weight: .semibold))
                    .foregroundStyle(Brand.accent)
                    .frame(width: size * 1.5, height: size * 1.5)
                    .background(Brand.accent.opacity(0.15), in: Circle())
            } else {
                ForEach(Array(shown.enumerated()), id: \.element.id) { index, member in
                    AvatarView(url: member.profile?.avatarUrl, name: member.name,
                               size: index == 0 ? size : size * 0.72)
                        .overlay(Circle().stroke(Brand.surface, lineWidth: 4))
                        .offset(Self.offsets[shown.count - 1][index].applying(
                            CGAffineTransform(scaleX: size / 64, y: size / 64)))
                }
            }
        }
    }

    /// Hand-placed so a cluster of any size reads as a group, not a row.
    private static let offsets: [[CGSize]] = [
        [CGSize(width: 0, height: 0)],
        [CGSize(width: -18, height: -6), CGSize(width: 26, height: 14)],
        [CGSize(width: -14, height: -12), CGSize(width: 32, height: -18), CGSize(width: 22, height: 30)],
        [CGSize(width: -16, height: -12), CGSize(width: 34, height: -22),
         CGSize(width: -30, height: 32), CGSize(width: 28, height: 30)],
    ]
}

private extension CGSize {
    func applying(_ t: CGAffineTransform) -> CGSize {
        CGSize(width: width * t.a, height: height * t.d)
    }
}

File: android/app/src/main/java/com/wsgpolar/disband/ui/calls/VoiceLobbySheet.kt
package com.wsgpolar.disband.ui.calls

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.MicOff
import androidx.compose.material.icons.filled.GraphicEq
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wsgpolar.disband.core.Brand
import com.wsgpolar.disband.core.Palette
import com.wsgpolar.disband.core.LocalPalette
import com.wsgpolar.disband.data.VoiceParticipant
import com.wsgpolar.disband.state.AppState
import com.wsgpolar.disband.ui.AvatarImage
import kotlinx.coroutines.launch

/**
 * Peek into a voice channel before joining: who's there, and one
 * big button to join. Muting first is a tap away, so nobody hears
 * you arrive.
 */
@Composable
fun VoiceLobbySheet(
    app: AppState,
    channelName: String,
    participants: List<VoiceParticipant>,
    onJoin: () -> Unit,
    onDismiss: () -> Unit,
) {
    val palette = LocalPalette.current
    val scope = rememberCoroutineScope()
    var joinMuted by remember { mutableStateOf(false) }
    var joining by remember { mutableStateOf(false) }

    Column(
        Modifier
            .fillMaxWidth()
            .background(palette.surface),
    ) {
        // Header
        Column(
            Modifier.padding(horizontal = 24.dp).padding(top = 28.dp, bottom = 16.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            AvatarCluster(members = participants, size = 64.dp)
            Spacer(Modifier.size(14.dp))
            Text(channelName, color = palette.textPrimary, fontSize = 22.sp, fontWeight = FontWeight.Bold)
            Text(
                memberSummary(participants),
                color = palette.textMuted, fontSize = 14.sp,
            )
        }

        // Member list
        LazyColumn(
            Modifier.weight(1f),
            contentPadding = PaddingValues(horizontal = 24.dp, vertical = 8.dp),
        ) {
            items(participants) { member ->
                MemberRow(member = member, palette = palette)
            }
        }

        // Controls
        Row(
            Modifier.fillMaxWidth()
                .padding(16.dp),
            horizontalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconButton(
                onClick = { joinMuted = !joinMuted },
                modifier = Modifier.size(58.dp).clip(CircleShape).background(if (joinMuted) Brand.dnd else palette.surfaceRaised),
            ) {
                Icon(
                    if (joinMuted) Icons.Filled.MicOff else Icons.Filled.Mic,
                    contentDescription = if (joinMuted) "Join muted" else "Join unmuted",
                    tint = if (joinMuted) Color.White else palette.textPrimary,
                    modifier = Modifier.size(22.dp),
                )
            }
            androidx.compose.material3.Button(
                onClick = {
                    joining = true
                    scope.launch {
                        onJoin()
                        joining = false
                    }
                },
                enabled = !joining,
                modifier = Modifier.weight(1f).height(58.dp),
                shape = CircleShape,
            ) {
                if (joining) {
                    androidx.compose.material3.CircularProgressIndicator(
                        color = Color.White, modifier = Modifier.size(22.dp),
                    )
                } else {
                    Icon(Icons.Filled.GraphicEq, contentDescription = null, tint = Color.White, modifier = Modifier.size(20.dp))
                    Spacer(Modifier.width(8.dp))
                    Text("Join Voice", color = Color.White, fontSize = 16.sp, fontWeight = FontWeight.Bold)
                }
            }
        }
    }
}

@Composable
private fun AvatarCluster(members: List<VoiceParticipant>, size: androidx.compose.ui.unit.Dp) {
    val palette = LocalPalette.current
    if (members.isEmpty()) {
        Box(
            Modifier.size(size * 1.5f).clip(CircleShape).background(palette.accent.copy(alpha = 0.15f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(Icons.Filled.GraphicEq, contentDescription = null, tint = palette.accent, modifier = Modifier.size(size * 0.5f))
        }
    } else {
        val shown = members.take(4)
        Box(Modifier.size(size * 1.5f)) {
            shown.forEachIndexed { index, member ->
                val offsetX = when (index) {
                    0 -> 0f
                    1 -> -18f
                    2 -> -14f
                    3 -> -16f
                    else -> 0f
                }
                val offsetY = when (index) {
                    0 -> 0f
                    1 -> -6f
                    2 -> -12f
                    3 -> -12f
                    else -> 0f
                }
                Box(
                    Modifier
                        .offset(offsetX.dp, offsetY.dp)
                        .size(if (index == 0) size else size * 0.72f),
                ) {
                    AvatarImage(url = member.profile?.avatarUrl, name = member.profile?.name ?: "Member", size = size)
                }
            }
        }
    }
}

@Composable
private fun MemberRow(member: VoiceParticipant, palette: Palette) {
    Row(
        Modifier.fillMaxWidth().padding(vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        AvatarImage(url = member.profile?.avatarUrl, name = member.profile?.name ?: "Member", size = 36.dp)
        Spacer(Modifier.width(12.dp))
        Text(member.profile?.name ?: "Member", color = palette.textPrimary, fontSize = 15.sp, fontWeight = FontWeight.Medium)
    }
}

private fun memberSummary(members: List<VoiceParticipant>): String {
    return when (members.size) {
        0 -> "No one's here yet — be the first."
        1 -> "${members[0].profile?.name ?: "Someone"} is hanging out"
        2 -> "${members[0].profile?.name} and ${members[1].profile?.name}"
        else -> "${members[0].profile?.name}, ${members[1].profile?.name} and ${members.size - 2} others"
    }
}
